tickets/DM-32079: Dilate child footprints using PSF footprint

fred3m opened the pull request:
Replace an incorrect thresholding algorithm with a more robust dilation algorithm to convolve scarlet models with the seeing in each band and properly truncate the resulting footprint.

Commit:
Remove HeavyFootprint test in test_scarlet

The test_scarlet.py module is for testing that the stack version
of scarlet can initialize sources with no errors. Any tests
for stack objects should be in test_deblend.py, so the test that
heavy footprints are made corectly was removed.

Commit:
Fix HeavyFootprints generated from scarlet models

`modelToHeavy` was modified to remove an incorrect thresholding
algorithm used to truncate convolved scarlet models and replace
it with an algorithm that grows the deconvolved footprint with
the rectangular PSF.

Commit:
Clip footprints to exposure

Deblended children near the edge of an image can have their
footprints dilated outside of the exposure by the PSF width.
This commit ensures that the dilated model is always contained
fully in the exposure. This is valid, as later in pipe_tasks
the science pipelines flag edge sources so it isn't expected
that all of their flux is contained in the image.

natelust commented on python/lsst/meas/extensions/scarlet/source.py:
```diff
@@ -56,34 +108,40 @@ def modelToHeavy(source, filters, xy0=Point2I(), observation=None, dtype=np.floa
-    mHeavy = afwDet.MultibandFootprint.fromArrays(filters, model, xy0=xy0)
-    peakCat = afwDet.PeakCatalog(source.detectedPeak.table)
+    # Get the SpanSet that contains all of the model pixels
+    model = np.array(source.get_model()).astype(np.float32)
```
dont do seperate `astype` here

Commit:
Revert to old modelToHeavy code with a slight change

The newer version of the code was creating the spans from the
scarlet model, dilating those spans, and using those spans
for the final footprints. The old version of the code simply
thresholded at zero to create the HeavyFootprints.

This commit reverts to the old code, with the change that it
performs the convolution and uses all pixels that are not zero
(including negative pixel values) in the footprint.

There is also a minor change to ensure that the MultibandFootprint
has all of the peaks, not just the single pand footprints.

Commit:
Clear scarlet cache to fix memory leak

Commit:
Remove unused import

Commit:
Use correct bounding box for MultibandFootprint

The original version of this code used the bounding box of the
spanset, not the mask that was used to create the spanset.
When the mask contains empty rows the two will differ, which
caused some sources to be placed incorrectly in the footprint.

Commit:
Rebase with master

Commit:
Remove unused import
